tools/fw-update/rs-fw-update.cpp

```diff
@@ -0,0 +1,272 @@
+    {
+        printf("\rfirmware update progress: %d[%%]", (int)(progress * 100));
+    });
+    std::cout << std::endl << std::endl << "firmware update done" << std::endl;
+}
+
+void list_devices(rs2::context ctx)
+{
+    auto devs = ctx.query_devices();
+    if (devs.size() == 0)
+    {
+        std::cout << std::endl << "there are no connected devices" << std::endl;
+        return;
+    }
+
+    std::cout << std::endl << "connected devices:" << std::endl;
+
+
+    int counter = 0;
+    for (auto&& d : devs)
+    {
+        std::cout << ++counter << ") ";
+        print_device_info(d);
+    }
+}
+
+int main(int argc, char** argv) try
+{
+    rs2::context ctx;
+
+    std::condition_variable cv;
+    std::mutex mutex;
+    std::string selected_serial_number;
+
+    rs2::device new_device;
+    rs2::update_device new_fw_update_device;
+
+    bool done = false;
+
+    CmdLine cmd("librealsense rs-fw-update tool", ' ', RS2_API_VERSION_STR);
+
+    SwitchArg list_devices_arg("l", "list_devices", "List all available devices");
+    SwitchArg recover_arg("r", "recover", "Recover all connected devices which are in recovery mode");
+    ValueArg<std::string> file_arg("f", "file", "Path of the firmware image file", false, "", "string");
+    ValueArg<std::string> serial_number_arg("s", "serial_number", "The serial number of the device to be update, this is mandetory if more than one device is connected", false, "", "string");
+
+    cmd.add(list_devices_arg);
+    cmd.add(recover_arg);
+    cmd.add(file_arg);
+    cmd.add(serial_number_arg);
+
+    cmd.parse(argc, argv);
+
+    bool recovery_request = recover_arg.getValue();
+
+    if (list_devices_arg.isSet())
+    {
+        list_devices(ctx);
+        return EXIT_SUCCESS;
+    }
+
+    if (serial_number_arg.isSet())
+    {
+        selected_serial_number = serial_number_arg.getValue();
+        std::cout << std::endl << "search for device with serial number: " << selected_serial_number << std::endl;
+    }
+
+    if (!file_arg.isSet())
+    {
+        std::cout << std::endl << "firmware file must be selected" << std::endl << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    std::string fw_file_path = file_arg.getValue();
+    std::vector<uint8_t> fw_image = read_fw_file(fw_file_path);
+
+    if (fw_image.size() == 0)
+    {
+        std::cout << std::endl << "failed to read firmware file" << std::endl << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    std::cout << std::endl << "update to FW: " << fw_file_path << std::endl;
+
+    std::string asic_serial_number;
+
+    // Recovery
+    bool recovery_executed = false;
+    if (recover_arg.isSet())
+    {
+        auto devs = ctx.query_devices(RS2_PRODUCT_LINE_DEPTH);
+        for (auto&& d : devs)
+        {
+            if (d.is<rs2::update_device>())
+            {
+                std::cout << std::endl << "recovering device: " << std::endl;
+                print_device_info(d);
+                update(d, fw_image);
```

Comment:
In case of multiple recovery devices, the same image file will be downloaded to both SR300 and D400

Reply:
I will add try-catch so all the devices that can be updated will be.